proto/engine/v1/electra.go

```diff
@@ -1,4 +1,121 @@
 package enginev1
 
+import (
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/pkg/errors"
+)
+
 type ExecutionPayloadElectra = ExecutionPayloadDeneb
 type ExecutionPayloadHeaderElectra = ExecutionPayloadHeaderDeneb
+
+var (
+	drExample = &DepositRequest{}
+	drSize    = drExample.SizeSSZ()
+	wrExample = &WithdrawalRequest{}
+	wrSize    = wrExample.SizeSSZ()
+	crExample = &ConsolidationRequest{}
+	crSize    = crExample.SizeSSZ()
+)
+
+const LenExecutionRequestsElectra = 3
+
+func (eee *ExecutionBundleElectra) GetDecodedExecutionRequests() (*ExecutionRequests, error) {
+	requests := &ExecutionRequests{}
+
+	if len(eee.ExecutionRequests) != LenExecutionRequestsElectra /* types of requests */ {
+		return nil, errors.Errorf("invalid execution request size: %d", len(eee.ExecutionRequests))
+	}
+
+	// deposit requests
+	drs, err := unmarshalItems(eee.ExecutionRequests[0], drSize, func() *DepositRequest { return &DepositRequest{} })
+	if err != nil {
+		return nil, err
+	}
+	requests.Deposits = drs
+
+	// withdrawal requests
+	wrs, err := unmarshalItems(eee.ExecutionRequests[1], wrSize, func() *WithdrawalRequest { return &WithdrawalRequest{} })
+	if err != nil {
+		return nil, err
+	}
+	requests.Withdrawals = wrs
+
+	// consolidation requests
+	crs, err := unmarshalItems(eee.ExecutionRequests[2], crSize, func() *ConsolidationRequest { return &ConsolidationRequest{} })
+	if err != nil {
+		return nil, err
+	}
+	requests.Consolidations = crs
+
+	return requests, nil
+}
+
+func EncodeExecutionRequests(requests *ExecutionRequests) ([]hexutil.Bytes, error) {
+	if requests == nil {
+		return nil, errors.New("invalid execution requests")
+	}
+
+	drBytes, err := marshalItems(requests.Deposits)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to marshal deposit requests")
+	}
+
+	wrBytes, err := marshalItems(requests.Withdrawals)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to marshal withdrawal requests")
+	}
+
+	crBytes, err := marshalItems(requests.Consolidations)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to marshal consolidation requests")
+	}
+	return []hexutil.Bytes{drBytes, wrBytes, crBytes}, nil
+}
+
+type sszUnmarshaler interface {
+	UnmarshalSSZ([]byte) error
+}
+
+type sszMarshaler interface {
+	MarshalSSZ() ([]byte, error)
+}
+
+func marshalItems[T sszMarshaler](items []T) ([]byte, error) {
+	var totalSize int
+	marshaledItems := make([][]byte, len(items))
+	for i, item := range items {
+		bytes, err := item.MarshalSSZ()
+		if err != nil {
+			return nil, fmt.Errorf("failed to marshal item at index %d: %w", i, err)
+		}
+		marshaledItems[i] = bytes
+		totalSize += len(bytes)
+	}
+
+	result := make([]byte, totalSize)
+
+	for i, bytes := range marshaledItems {
+		copy(result[i*len(bytes):(i+1)*len(bytes)], bytes)
```

Comment:
Use the `MarshalSSZTo` method instead to avoid the extra copy. Like this:
```
type sszMarshaler interface {
	MarshalSSZ() ([]byte, error)
	SizeSSZ() int
}

func marshalItems[T sszMarshaler](items []T) ([]byte, error) {
        if len(items) == 0 {
                return []byte{}, nil
        }
        size := items[0].SizeSSZ()
        buf := make([]byte, size * len(items))
	for i, item := range items {
		if err := item.MarshalSSZTo(buf[i*size:(i+1)*size]); err != nil {
		        return nil, fmt.Errorf("failed to marshal item at index %d: %w", i, err)
		}
	}
	return buf, nil
}
```

Reply:
thanks

Reply:
actually not sure if it will work as cleanly as this, will ask on slack

Reply:
fixed